client/task_runner_test.go

```diff
@@ -195,3 +198,65 @@ func TestTaskRunner_SaveRestoreState(t *testing.T) {
 	}
 }
 */
+
+func TestTaskRunner_HandleDiscovery(t *testing.T) {
+	// Create the task runner
+	_, tr := testTaskRunner()
+
+	// Modify the job so we have predictable results
+	tr.allocID = "1"
+	tr.jobID = "job1"
+	tr.taskGroup = "group1"
+	tr.task.Name = "web"
+
+	// Assign dynamic ports. The ReservedPorts are overloaded for dynamic
+	// ports also, so we set those here as well.
+	tr.task.Resources.Networks[0].DynamicPorts = []string{"http", "https"}
+	tr.task.Resources.Networks[0].ReservedPorts = []int{80, 443}
+
+	// Create the mock discovery layer
+	providers := []discovery.Factory{discovery.NewMockDiscovery}
+	disc, err := discovery.NewDiscoveryLayer(providers, tr.config, tr.logger, tr.config.Node)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	tr.discovery = disc
+	mp := disc.Providers[0].(*discovery.MockDiscovery)
+
+	// Register the task with discovery
+	tr.handleDiscovery(false)
+
+	// Check the result
+	expect := map[string]int{
+		"job1.group1.web:1":       0,
+		"job1.group1.web.http:1":  80,
+		"job1.group1.web.https:1": 443,
```

Comment:
> Lets say I have my redis example container spun up with docker host network mode.
The redis server will be listening at 0.0.0.0:6379 on a HostOS that can have a single or multiple IP addresses. 

Ah, I see. Thanks for clarifying!

When nomad schedules a task it creates a port allocation on the host on only one of the interfaces, so internally we have an IP. You have some control over the allocation via subnet filter (though you can't choose the exact IP).

Last time I looked at the network resource bid code, it does not bid for multiple interfaces so binding to 0.0.0.0 is probably not supported by the scheduler (though maybe it's been updated). I will need to review this.

I think our options are to disallow host mode and limit to individual interfaces, or make the allocation reserve the same port on all interfaces. I suspect right now we're subject to collisions if host mode networking is used.